### ISSUE TITLE
Camera boundry

### ISSUE THREAD

**Issue by hpinkos** [AnalyticalGraphicsInc/cesium#4802]
I've seen this requested on the forum a handful of times, most recently here: https://groups.google.com/forum/?hl=en#!topic/cesium-dev/z-kWiAZ1Jfg

For some applications, it would be helpful to define a rectangle to restrict the camera to view only a certain span of the map.

**Comment by mramato** [comment]
I've seen other globes (at least WorldWind I think) handle this using a separate class, we could call ours `CameraLimiter` or something similar.  This class would allow you to set min/max values for anything that makes sense  on the Camera.  Then every frame, it's job is to check for out of bounds values and set them to the min/max if they are outside the desired configuration.  This should provide a flexible and elegant solution and allow us to set limits on all kinds of things (boundary, rotation, direction, etc..)

**Comment by hpinkos** [comment]
Also reported here: https://groups.google.com/forum/?hl=en#!topic/cesium-dev/x8-uWKKlYNc

**Comment by rahwang** [comment]
Also here: https://groups.google.com/d/msg/cesium-dev/tAp48bXcQRw/NmTXbO90AwAJ

**Comment by pjcozzi** [comment]
@rahwang this could be a good one for you for next month if you are interested.

**Comment by rahwang** [comment]
Yes please! I'll add this to my list

**Comment by hpinkos** [comment]
This will be helpful for something I'm working on too =)

**Comment by mramato** [comment]
Yep, I would love to see this because we'll have good use for it on cesium.com, has been on my todo list forever.  I've has discussions with bagnell in the past on how best to implement this, so let me know if you want to talk before tackling it.

**Comment by pjcozzi** [comment]
CC #3111

**Comment by pjcozzi** [comment]
Sketchfab just added a similar feature: https://blog.sketchfab.com/introducing-camera-constraints/

**Comment by pizza2code** [comment]
I would appreciate this as well!  Even a simple boolean option such as "clampCameraToEdges" for the 2D map would be great.  It seems of marginal utility to let people scroll past the north & south poles, as you currently can.

**Comment by hpinkos** [comment]
Someone requested this at the FOSS4G cesium workshop as well.  I think this would be a really popular feature if anyone has the chance to look at it soon.

**Comment by pjcozzi** [comment]
It's also a nice onramping issue.  @ggetz please add this to your backlog.

**Comment by scottnc27603** [comment]
Any updates on when this feature will make it into a release?

**Comment by pjcozzi** [comment]
@scottnc27603 no update, but contributions are very welcome if you have the bandwidth.

**Comment by OmarShehata** [comment]
Requested on the forum again: https://groups.google.com/forum/#!topic/cesium-dev/NzcIdGyfSt4

**Comment by mramato** [comment]
@OmarShehata also had some thoughts in https://github.com/AnalyticalGraphicsInc/cesium/issues/8143#issue-491026306

**Comment by GuniBMB** [comment]
where can i get the imp code?

**Comment by ggetz** [comment]
Requested in the forum: https://community.cesium.com/t/is-there-any-way-to-constrain-extent/26745/7

**Comment by ggetz** [comment]
Requested on the forum: https://community.cesium.com/t/limit-user-camera-movement/22992/3

**Comment by groundrace** [comment]
Hello,
is there any update on this feature?
I was able to limit the camera min/max elevation using the: `Cesium.ScreenSpaceCameraController.minimumZoomDistance`/`Cesium.ScreenSpaceCameraController.maximumZoomDistance` but i can't find anything in terms of bounding box where the camera should stay

**Comment by Spinksy** [comment]
> Hello, is there any update on this feature? I was able to limit (and it works pretty well!) the camera min/max elevation using the: `Cesium.ScreenSpaceCameraController.minimumZoomDistance`/`Cesium.ScreenSpaceCameraController.maximumZoomDistance` but i can't find anything in terms of bounding box where the camera should stay

Would also be interested to hear if there are any plans to implement this feature.

**Comment by romain974** [comment]
I have a working cesium class that allow to limit camera movement by bbox + zmax, or by entity + distance in meter. Contact me if you are interested.

**Comment by mattpgreen** [comment]
> I have a working cesium class that allow to limit camera movement by bbox + zmax, or by entity + distance in meter. Contact me if you are interested.

Hi Romian - could you please guide on how to limit camera movement? Thanks, Matt

**Comment by mattpgreen** [comment]
Hi @romain974 - could you please guide on how to limit camera movement? Thanks, Matt

**Comment by jjspace** [comment]
Requested on the forum https://community.cesium.com/t/limit-camera-to-a-zone-volume/34347

**Comment by mattpgreen** [comment]
Hi @romain974 - could you please guide on how to limit camera movement? Thanks, Matt
Hi @jjspace - could you please guide on this and help @adcomere on #34347
Hi @pjcozzi - could you please help get this function in real soon so that so many local community virtual twins can be implemented and kept will the local focus that's important to them? - thanks, Matt

**Comment by romain974** [comment]
One way to do this by yourself  is to use scene postrender event to test camera distance // limit and allow or stop camera move.
You also need to store each previous camera value to be able to restore it when camera reach limits.
If you wan't a full working code, contact me to purchase the module.

**Comment by DoisKoh** [comment]
I'm also adding more noise here to express our desire of this feature.

Thank you.

**Comment by romain974** [comment]
I need some time to generate a public, open source code.
=> JS Class with functions to limit camera by bounding box or distance // cartesian3  +  signal when camera hits limits

 If you want to contribute to this project:  https://github.com/sponsors/romain974?frequency=one-time

Target => mid october

**Comment by andrew-hossack** [comment]
Bump. Need a way to do this as well!

**Comment by romain974** [comment]
Hi,

Thank you for reaching out! You can find the source code in my open-source repository here: [https://github.com/romain974/cesiumCameraLimiter](https://github.com/romain974/cesiumCameraLimiter).

Feel free to let me know if you encounter any bugs or have suggestions for improvements.

If you find the project useful, I'd greatly appreciate it if you could support it through my [GitHub Sponsors page](https://github.com/sponsors/romain974). Your contributions help me continue improving the project.

Best regards,
